sktree/stats/forestht.py

```diff
@@ -0,0 +1,794 @@
+        The indices of the samples used in the training set.
+
+    indices_test_ : ArrayLike of shape (n_samples_test,)
+        The indices of the samples used in the testing set.
+
+    samples_ : ArrayLike of shape (n_samples_final,)
+        The indices of the samples used in the final test set that would slice
+        the original ``(X, y)`` input.
+
+    y_true_final_ : ArrayLike of shape (n_samples_final,)
+        The true labels of the samples used in the final test.
+
+    posterior_final_ : ArrayLike of shape (n_samples_final,)
+        The predicted posterior probabilities of the samples used in the final test.
+
+    null_dist_ : ArrayLike of shape (n_repeats,)
+        The null distribution of the test statistic.
+
+    Notes
+    -----
+    This class trains two forests: one on the original dataset, and one on the
+    permuted dataset. The forest from the original dataset is cached and re-used to
+    compute the test-statistic each time the :meth:`test` method is called. However,
+    the forest from the permuted dataset is re-trained each time the :meth:`test` is called
+    if the ``covariate_index`` differs from the previous run.
+
+    To fully start from a new dataset, call the ``reset`` method, which will then
+    re-train both forests upon calling the :meth:`test` and :meth:`statistic` methods.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        estimator=None,
+        random_state=None,
+        verbose=0,
+        test_size=0.2,
+        permute_per_tree=True,
+        sample_dataset_per_tree=True,
+    ):
+        super().__init__(
+            estimator=estimator,
+            random_state=random_state,
+            verbose=verbose,
+            test_size=test_size,
+            permute_per_tree=permute_per_tree,
+            sample_dataset_per_tree=sample_dataset_per_tree,
+        )
+
+    def _get_estimator(self):
+        if self.estimator is None:
+            estimator_ = RandomForestClassifier()
+        elif not isinstance(self.estimator, (ForestClassifier, sklearnForestClassifier)):
+            raise RuntimeError(f"Estimator must be a ForestClassifier, got {type(self.estimator)}")
+        else:
+            # self.estimator is an instance of a ForestEstimator
+            estimator_ = self.estimator
+        return estimator_
+
+    def _statistic(
+        self,
+        estimator: ForestClassifier,
+        X: ArrayLike,
+        y: ArrayLike,
+        covariate_index: ArrayLike = None,
+        metric="mi",
+        return_posteriors: bool = False,
+        **metric_kwargs,
+    ):
+        """Helper function to compute the test statistic."""
+        metric_func = METRIC_FUNCTIONS[metric]
+        rng = np.random.default_rng(self.random_state)
+
+        if metric in POSTERIOR_FUNCTIONS:
+            predict_posteriors = True
+        else:
+            predict_posteriors = False
+
+        if self.permute_per_tree:
```

Comment:
Same as comment above, should be `if self.sample_dataset_per_tree`

Reply:
Resolved.